### ISSUE TITLE
UNSUBSCRIBED.Request: session scope or global scope?

### ISSUE THREAD

**Issue by ggs-sglimm** [tavendo/WAMP#161]
Building Blocks|Identifiers|IDs states that UNSUBSCRIBE.Request is a Session Scope ID, and should "be incremented by 1 beginning with 1"

Publish and Subscribe|Subscribing and Unsubscribing|UNSUBSCRIBE states that "Request is a random, ephemeral ID chosen by the Subscriber"

Which of these two is true?

**Comment by mulderr** [comment]
Same problem with SUBSCRIBE.Request.

Interestingly browsing through [AutobahnJS code](https://github.com/tavendo/AutobahnJS/blob/master/package/lib/session.js#L62) we can see that it uses a *single* function for generating all request IDs and makes them all random:

``` javascript
// generate a WAMP ID
//
function newid () {
   return Math.floor(Math.random() * 9007199254740992);
}
```

I guess the intention of the spec was to make it a recommendation (SHOULD) that the ID is *"incremented by 1 beginning with 1"* but allow implementations to divert. In which case the the description of *Publish and Subscribe|Subscribing and Unsubscribing|UNSUBSCRIBE* (amongst others) needs to be changed and *"Request is a random, ephemeral ID chosen by the Subscriber"* is incorrectly too strict.

Question is then do we really need the session scope recommendation in the first place if we are going to allow different behavior everywhere. Why not just say the implementaion is free to choose a way generate session IDs as long as the likeklyhood of duplicates is no worse than when using a well behaved pseudo random generator with uniform distribution over [0, 2^53]. Which is not the same as saying they must be random.

**Comment by mulderr** [comment]
FYI [AutobahnCpp](https://github.com/tavendo/AutobahnCpp/blob/master/autobahn/autobahn.hpp#L421) goes the opposite way. It implements the recommendation and stores a counter for each session inrementing it for each request. No randomness. 

Thus it ignores *"Request is a random, ephemeral ID chosen by the Subscriber"*.

Aha! AutobahnCpp does not fully comply with the Basic Profile! :P

This really could use a clarification. Either way.

**Comment by goeddea** [comment]
Will ask Tobias, but AFAIK the "random, ephemeral ID" is the way to go.

For scenarios with a single router, @mulderr is correct that the generation method does not matter as long as the likelihood of duplicates is low enough. 

Once you get to multi-router scenarios, you do not want to have to coordinate between routers for request IDs etc, so random is the way to go there.

**Comment by meejah** [comment]
It's probably worth noting that if the IDs are random the range needs to be enormous or else (due to birthday problem) if there are a large number of outstanding requests the IDs can overlap. @oberstet and I debugged a problem like this in crossbar when load testing.

**Comment by mulderr** [comment]
It does not need a large number of outstanding requests when the IDs are sequential like they are in crossbar now. It needs one request sitting there for a really long time. Eventually the IDs will wrap around and it will get overwritten because crossbar does not check if the generated ID is not already being used. See [here](https://github.com/crossbario/crossbar/blob/master/crossbar/router/dealer.py#L417) for invocations. But I'm not sure how that came up during load testing since it would take 

    > 2^53 / 10^6 / 60 / 60 / 24 / 365.25
    285.42092094268867

this many years at a million requests per second.

Lets focus on random IDs though because they seem to be the way this is going. We calculate the collision probability for `10^n` WAMP global IDs.

n | 10^n | P(collision) | what does it mean?
---|------|------------|---
2 | 100 | 5.5e-13 | pretty ok, very close to the probability of a duplicate for 2^41 (!) UUIDs
3 | 1k | 5.5e-11   | still cool, about 2 orders of magnitude lower than the probability of a cosmic ray induced failure per byte per month (yay for ECC!)
4 | 10k | 5.6e-9   | about 10 times lower than the chance of winning a lottery
5 | 100k | 5.6e-7 | lower (but not by much!) than the chance of getting hit (not necesserily killed) by lightning in any one year (for U.S. citizens)
6 | 1M | 5.6e-5    | breached the proverbial 1 in a million
7 | 10M | 5.5e-3  | 0.5%? we passed the "this system is broken by design" point of no return somewhere above
8 | 100M | 0.43   | almost a coin toss

At 100k we have 5.6e-7 probability of ID collision. If you think 100k is unrealisitic think 10k sessions with 10 outstanding invocations each. That probability is too high. Especially so if you add time to the equation by allowing a given scenario to reapeat over and over again. 

The conclusion is that with the space of only 2^53 possible values we cannot automatically rely on random IDs being unique. A router must check if a generated ID is not already being used or else it may overwrite another request.

Clients should also take this into account but it's much better in that case. I'm not sure it's likely a single client would have tens of thousands of outstanding requests of a given type at a time.

----

Further lets consider multi-router scenarios since they already got mentioned in this context. I'm not really sure how a multi-router setup would work but let's see how much of a problem this limited range could really be by focusing on session IDs under the assumption that they should be unique for a set of routers.

We assume all routers generate session IDs according to specification: randomly with uniform distrubution over [0, 2^53]. Note it is *not* a problem if a router generates a session ID that collides with a session it is currently serving itself. It can easily detect this and generate another session ID until it finds an unused one - as described above. The real problem occurs when we add more routers and would like session IDs to be unique across the whole set. Like @goeddea mentioned we would like to avoid having to coordinate between routers. But then it is up to us to ensure the collision probability is at an acceptable level - whatever that means.

We calculte the session ID collision probability `p` for `r` routers, each serving `n` sessions. If I am correct this will be exactly... a formula that would be totally unreadable without proper formatting. So let's just consider a couple cases again.

n | p r=2 | p r=3 | p r=10
---|---|---|---
1k | 1.7e-10 | 4.4e-10 | 5.5e-9
10k | 1.7e-8 | 4.4e-8 | 5.5e-7
100k | 1.7e-6 | 4.4e-6 | 5.5e-5
1M | 1.7e-4 | 4.4e-4 | 5.5e-3

Cool. We have some numbers. What would those probabilities be if we repeat everything t times? We already have t=1 above. Now t=1k

n | p r=2 | p r=3 | p r=10
---|---|---|---
1k | 1.7e-7 | 4.4e-7 | 5.5e-6
10k | 1.7e-5 | 4.4e-5 | 5.5e-4
100k | 1.7e-3 | 4.4e-3 | 0.05
1M | 0.16 | 0.36 | 0.99

In short skipping coordination in this scenario is not a viable option unless you expect to have relatively few sessions per router which begs the question why would you go multi-router in the first place (really, why would you?). 2^53 possible values is just not enough. Note that UUIDv4s which would be probably considered suitable here have 2^122 possible values which doesn't even begin to compare.

One possible solution could be to pre-configure a distinct range of IDs for each router and draw randomly from that.

And thats for session IDs. Relying on request IDs being unique for a set of routers would be just wrong - see above, it does not even realistically hold for a single router!

---

Thoughts?

**Comment by meejah** [comment]
I think the aproximation for a collision is 2^(n/2)? I can look more closely at this later on but it should be sufficient to consider only the total number of active sessions (no matter how many routers they're spread over) so random uniform draws from 2^53 IDs should allow for around 2^26.5 active sessions before a collision is expected (so close to 100M sessions).

I believe the reason 2^53 was chosen is because of limits in javascript -- although a little more breathing room might be nice ;)

At one point req IDs for something used 2^31 as a range (left over from a PR that wanted to use signed 32 bit int for the req id) so collisions happened with far fewer outstanding requests in that case ;)

**Comment by mulderr** [comment]
> I think the aproximation for a collision is 2^(n/2)?

The numbers are low enough that you can actually calcuate this exactly. The collision probability `p(c)=1 - p(nc)` where `p(nc)` is the probability that there where *no* collisions.

```
        N N-1 N-2     N-(n-1)
p(nc) = - --- --- ... -------
        N  N   N         N
```

where `N=2^53`. First ID has `N` possibilities, second `N-1`. In general n-th ID is left with `N-(n-1)` possible values. It's a bit more complicated with multiple routers but follows the same logic...

> I can look more closely at this later on but it should be sufficient to consider only the total number of active sessions (no matter how many routers they're spread over)

That was my initial thought but each router can easily detect a collision as long as the ID collides with a local session, so I didn't want to count those because they can be trivially avoided.

> I believe the reason 2^53 was chosen is because of limits in javascript

Yes, javascript unfortunatelly stores everything as double so integers outside (-2^53,2^53) cannot be properly represented. Indeed a bit more breathing room would be nice. Maybe a thing to keep in mind for v3? :)

**Comment by meejah** [comment]
With the javascript constraint, there'd have to be some sort of terrible hack like splitting the ID into two numbers or similar (i.e. giving 2^106). "Or something".

Making the session-IDs sequential (and therefore predictable) would at a bare minimum leak information about how many sessions there are (and perhaps other bad things I'm not thinking of right now).

**Comment by mulderr** [comment]
@meejah agreed, two integers would be a terrible hack

WAMP wants to be language agnostic and it's not possible to rely on various languages providing integer types with high enough bounds with any consistency, so it may be necessary to eventually ditch the integer idea altogether. Is it even worthwhile to consider moving to byte arrays or so at this point?

**Comment by meejah** [comment]
Maybe 12 digits of base64-encoded data (2^72)?

**Comment by mulderr** [comment]
Ok, so if we are considering this, IMHO it would be best to follow the established practice of at least ~128 bits for globally unique IDs. IPv6 addresses and UUIDv4s are both 128 bit (minus the 6 metadata bits for UUIDv4). It does not have to be UUIDv4 in particular but why reinvent the wheel when we can possibly use existing libraries.

It would be most efficient to store it as a byte array. [Uint8Array](https://developer.mozilla.org/en/docs/Web/JavaScript/Typed_arrays) seems to be a good fit for JS and as you mentioned it could be base64 encoded during serialization to JSON. Don't know about MsgPack but I suppose it would be even simpler.
